src/Storages/StorageReplicatedMergeTree.cpp

```diff
@@ -5373,41 +5375,55 @@ void StorageReplicatedMergeTree::fetchPartition(
     if (from.empty())
         throw Exception("ZooKeeper path should not be empty", ErrorCodes::ILLEGAL_TYPE_OF_ARGUMENT);
 
-    String partition_id = getPartitionIDFromQuery(partition, query_context);
     zkutil::ZooKeeperPtr zookeeper;
     if (auxiliary_zookeeper_name != default_zookeeper_name)
-    {
         zookeeper = getContext()->getAuxiliaryZooKeeper(auxiliary_zookeeper_name);
-
-        LOG_INFO(log, "Will fetch partition {} from shard {} (auxiliary zookeeper '{}')", partition_id, from_, auxiliary_zookeeper_name);
-    }
     else
-    {
         zookeeper = getZooKeeper();
 
-        LOG_INFO(log, "Will fetch partition {} from shard {}", partition_id, from_);
-    }
-
     if (from.back() == '/')
         from.resize(from.size() - 1);
 
-
-    /** Let's check that there is no such partition in the `detached` directory (where we will write the downloaded parts).
-      * Unreliable (there is a race condition) - such a partition may appear a little later.
-      */
-    Poco::DirectoryIterator dir_end;
-    for (const std::string & path : getDataPaths())
+    if (fetch_part)
     {
-        for (Poco::DirectoryIterator dir_it{path + "detached/"}; dir_it != dir_end; ++dir_it)
+        String part_name = partition->as<ASTLiteral &>().value.safeGet<String>();
+        auto part_path = findReplicaHavingPart(part_name, from, zookeeper);
+
+        if (part_path.empty())
+            throw Exception("fetch part " + part_name + " not exists !", ErrorCodes::PART_DOESNT_EXIST);
```

Comment:
```suggestion
            throw Exception(ErrorCodes::PART_DOESNT_EXIST, "Part {} does not exist on any replica", part_name);
```

Reply:
i see many throw Exception at StorageReplicatedMergeTreep.cpp put the error code at the last

Reply:
Yes, you can throw exceptions this way, but the comment is more about exception message. It should start with capital letter and should not contain exclamation points.

Reply:
3ks,done